xwiki-rendering-wikimodel/src/main/javacc/XWiki20Scanner.jj

```diff
@@ -422,8 +422,13 @@ TOKEN_MGR_DECLS: {
     | <#MACRO_NAME: (<CHAR>)+ (["-", "_", ".", ":"] (<CHAR>)+)* >
     | <#MACRO_PARAMS: ( <DOUBLE_QUOTED> | <SINGLE_QUOTED> | "~" ~[] | ~["}", "\"", "'", "~"] | "}" ~["}", "~"] | "}~" ~[] )* >
     | <#HEADER_BEGIN_PATTERN: (<SPACE>)* ("=")+ (<SPACE>)* >
-    | <#REFERENCE_IMAGE: "[[image:" ("~" ~[] | ~["]", "~"] | "]" ~["]", "~"]  | "]~" ~[] )* "]]" >
-    | <#REFERENCE:  ( "[[" ( <REFERENCE_IMAGE> | "~" ~[] | ~["]", "~"] | "]" ~["]", "~"] | "]~" ~[] )* "]]" ) >
+    | <#REFERENCE_INNER: "~" ~[] | ~["]", "~"] | "]" ~["]", "~"] | "]~" ~[] >
+    | <#REFERENCE_IMAGE_INNER: "image:" (<REFERENCE_INNER>)* >
+    | <#REFERENCE_IMAGE: "[[" <REFERENCE_IMAGE_INNER> "]]" >
+    | <#REFERENCE_SINGLE:  ( "[[" ( <REFERENCE_IMAGE> | <REFERENCE_INNER> )* "]]" ) >
+    // Allow nested references, but only if the reference ends with an image reference
+    | <#REFERENCE:  ( "[[" ( <REFERENCE_IMAGE> | <REFERENCE_INNER> )* (( <REFERENCE_SINGLE> | <REFERENCE_IMAGE> |
+    <REFERENCE_INNER> )* ">>" <REFERENCE_IMAGE_INNER>)? "]]" ) >
```

Comment:
Modifying the behavior of the low level parser might be a risk in terms of retro compatibility which usually leads to introducing a new syntax (2.2) but it's not fully clear to me what can be the actual consequence here. Do you have some example ?